cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go

```diff
@@ -0,0 +1,114 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterapi
+
+import (
+	"path"
+	"reflect"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	kubeinformers "k8s.io/client-go/informers"
+	fakekube "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/kubernetes/pkg/controller"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	fakeclusterapi "sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset/fake"
+	informers "sigs.k8s.io/cluster-api/pkg/client/informers_generated/externalversions"
+)
+
+func mustCreateTestController(t *testing.T) *machineController {
+	t.Helper()
+
+	kubeclientSet := fakekube.NewSimpleClientset()
+	clusterclientSet := fakeclusterapi.NewSimpleClientset()
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeclientSet, controller.NoResyncPeriodFunc())
+	clusterInformerFactory := informers.NewSharedInformerFactory(clusterclientSet, controller.NoResyncPeriodFunc())
+
+	controller, err := newMachineController(kubeInformerFactory, clusterInformerFactory)
+	if err != nil {
+		t.Fatalf("failed to create test controller")
+	}
+
+	if err := controller.run(make(chan struct{})); err != nil {
+		t.Fatalf("failed to run controller: %v", err)
+	}
+
+	return controller
+}
+
+func TestFindMachineByID(t *testing.T) {
+	controller := mustCreateTestController(t)
+
+	testMachine := &v1alpha1.Machine{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "test",
+			Namespace: "test-namespace",
+		},
+	}
+
+	// Verify machine count starts at 0.
+	machines, err := controller.machineInformer.Lister().Machines(v1.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		t.Fatalf("unexpected error, got %v", err)
+	}
+	if len(machines) != 0 {
+		t.Fatalf("expected 0 machines, got %d", len(machines))
+	}
+
+	controller.machineInformer.Informer().GetStore().Add(testMachine)
+
+	// Verify machine count goes to 1.
+	machines, err = controller.machineInformer.Lister().Machines(v1.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		t.Fatalf("unexpected error, got %v", err)
+	}
+	if len(machines) != 1 {
+		t.Fatalf("expected 1 machine, got %d", len(machines))
+	}
+
+	// Verify inserted machine matches retrieved machine
+	if !reflect.DeepEqual(*machines[0], *testMachine) {
+		t.Fatalf("expected machines to be equal")
+	}
+
+	// Verify findMachine() can find the test machine
+	foundMachine, err := controller.findMachine(path.Join(testMachine.Namespace, testMachine.Name))
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	if foundMachine == nil {
+		t.Fatalf("expected to find machine %q in namespace %q", testMachine.Name, testMachine.Namespace)
+	}
+
+	// Verify that a successful findMachine returns a DeepCopy().
+	if foundMachine == testMachine {
```

Comment:
`reflect.DeepEqual` instead of `==`

Reply:
I want to test the pointer values are different.